File: ch03/labeling/meta_labeling.py
import numpy as np
import pandas as pd
import sys
import os

# Add project root to path so we can import utils
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..')))
from utils.multiprocess import mp_pandas_obj
from .triple_barrier import apply_pt_sl_on_t1

# Meta-Labeling — AFML Chapter 3, Sections 3.6-3.9, pages 50-54
#
# Meta-labeling is a two-stage approach to building trading strategies:
#
# Stage 1 — Primary model:
#   A simple model (or rule) that decides the SIDE of each trade (+1 = long, -1 = short).
#   It doesn't need to be highly accurate — it just needs to identify the direction.
#
# Stage 2 — Meta-model (meta-labeling):
#   A second model that looks at each primary signal and predicts whether to ACT on it.
#   It outputs a position size between 0 and 1:
#     0 = don't trade (primary model is probably wrong here)
#     1 = trade at full size (primary model is probably right here)
#
# --- Why does this work better than one model? ---
# Separating the "which direction" and "whether to trade" decisions allows each model
# to specialize. The meta-model learns to recognize when the primary model's signals
# are reliable vs when they should be ignored.
#
# --- How labels change with meta-labeling ---
# Without meta-labeling: bin ∈ {-1, 0, 1} — direction of price move
# With meta-labeling:    bin ∈ {0, 1}     — did the primary model's bet pay off?
#
# This turns the problem into a binary classification — much easier for ML models.


def get_events_meta(close, t_events, pt_sl, trgt, min_ret, num_threads=1, t1=False, side=None):
    # Expanded getEvents with Meta-Labeling — AFML Chapter 3, Snippet 3.6, page 51
    #
    # Extension of get_events() from triple_barrier.py that adds support for
    # a primary model's side predictions. When side is provided, the barriers
    # become asymmetric — profit target and stop loss are applied relative to
    # the predicted direction rather than always assuming a long position.
    #
    # Uses mpPandasObj to parallelize apply_pt_sl_on_t1 across CPU cores,
    # identical to how get_events() does it in triple_barrier.py.
    #
    # --- What changes with meta-labeling ---
    # Without side: barriers are symmetric (same width up and down, assume long)
    # With side:    barriers follow the primary model's direction.
    #               If primary says long (+1): upper = profit, lower = stop loss
    #               If primary says short (-1): upper = stop loss, lower = profit
    #
    # --- Inputs ---
    # close       : pd.Series — closing prices indexed by datetime
    # t_events    : pd.DatetimeIndex — candidate entry dates (from CUSUM filter)
    # pt_sl       : list [pt_multiplier, sl_multiplier]
    # trgt        : pd.Series — barrier width per event (typically daily volatility)
    # min_ret     : float — minimum target return to include an event
    # num_threads : int — number of parallel workers (default 1)
    # t1          : pd.Series or False — vertical barrier timestamps
    # side        : pd.Series or None — primary model's predicted side per event
    #               +1 = primary model predicts price will rise (long)
    #               -1 = primary model predicts price will fall (short)
    #               None = no primary model, use symmetric barriers (same as get_events)
    #
    # --- Output ---
    # pd.DataFrame with columns:
    #   t1   : timestamp of first barrier touch
    #   trgt : barrier width used
    #   side : primary model's side (only present if side was provided)

    # Step 1: Filter trgt to event dates and apply minimum return filter
    trgt = trgt.reindex(t_events, method='bfill')
    trgt = trgt[trgt > min_ret]

    # Step 2: Set up vertical barriers
    if t1 is False:
        t1 = pd.Series(pd.NaT, index=t_events)

    # Step 3: Determine side and barrier multipliers
    # If side is None: assume long (+1) for all events, symmetric barriers
    # If side is provided: use primary model's predictions, may be asymmetric
    if side is None:
        # No primary model — assume long, use same multiplier for both barriers
        side_  = pd.Series(1., index=trgt.index)
        pt_sl_ = [pt_sl[0], pt_sl[0]]  # symmetric: same multiplier both sides
    else:
        # Primary model provided — use its side predictions
        side_  = side.loc[trgt.index]  # align side to filtered event dates
        pt_sl_ = pt_sl[:2]             # [profit_target_mult, stop_loss_mult]

    # Step 4: Build events DataFrame
    events = pd.concat(
        {'t1': t1, 'trgt': trgt, 'side': side_},
        axis=1
    ).dropna(subset=['trgt'])

    # Step 5: Find first barrier touch using mpPandasObj
    # Identical pattern to get_events() — parallelizes apply_pt_sl_on_t1
    # across events.index using num_threads worker processes.
    df0 = mp_pandas_obj(
        func=apply_pt_sl_on_t1,
        pd_obj=('molecule', events.index),
        num_threads=num_threads,
        close=close,
        events=events,
        pt_sl=pt_sl_
    )

    # Step 6: Take earliest touch across all barriers
    events['t1'] = df0.dropna(how='all').min(axis=1)

    # Drop side column if no primary model was used
    if side is None:
        events = events.drop('side', axis=1)

    return events


def get_bins_meta(events, close):
    # Expanded getBins with Meta-Labeling — AFML Chapter 3, Snippet 3.7, pages 51-52
    #
    # Extension of get_bins() from triple_barrier.py that handles meta-labeling.
    # The key difference is how labels are assigned when a primary model's side
    # is available:
    #
    # Case 1 — No side (standard triple barrier):
    #   bin = sign(return)
    #   bin ∈ {-1, 0, 1} — did price go up, stay flat, or go down?
    #
    # Case 2 — With side (meta-labeling):
    #   bin = 1 if the primary model's bet was correct (made money)
    #   bin = 0 if the primary model's bet was wrong (lost money)
    #   bin ∈ {0, 1} — was the primary model RIGHT or WRONG?
    #
    # This converts the problem to binary classification — much easier for ML.
    #
    # --- Inputs ---
    # events : pd.DataFrame — output of get_events_meta()
    # close  : pd.Series — closing prices indexed by datetime
    #
    # --- Output ---
    # pd.DataFrame with columns:
    #   ret : actual return from entry to first barrier touch
    #   bin : label (see Cases above)

    # Step 1: Drop events with no barrier touch
    events_ = events.dropna(subset=['t1'])

    # Step 2: Build aligned price series covering entry and exit dates
    px = events_.index.union(events_['t1'].values).drop_duplicates()
    px = close.reindex(px, method='bfill')

    # Step 3: Compute returns from entry to first barrier touch
    out = pd.DataFrame(index=events_.index)

    exit_prices = []
    for t1 in events_['t1'].values:
        idx = close.index.searchsorted(t1)
        idx = min(idx, len(close) - 1)
        exit_prices.append(close.iloc[idx])

    entry_prices = close.reindex(events_.index, method='bfill').values
    out['ret'] = (np.array(exit_prices) / entry_prices) - 1

    # Step 4: Assign labels
    if 'side' in events_:
        # Meta-labeling: multiply return by primary model's side
        # Long (+1) and price up  → positive → correct → bin=1
        # Short (-1) and price down → negative × -1 → positive → bin=1
        # Wrong direction → negative → bin=0
        out['ret'] = out['ret'] * events_['side']
        out['bin'] = np.sign(out['ret'])
        out.loc[out['ret'] <= 0, 'bin'] = 0   # neutral counts as wrong
    else:
        # Standard: label by direction of price move
        out['bin'] = np.sign(out['ret'])

    return out


def drop_labels(events, min_pct=0.05):
    # Dropping Under-Populated Labels — AFML Chapter 3, Snippet 3.8, page 53
    #
    # Recursively removes observations with extremely rare labels until either:
    #   - All remaining labels appear in at least min_pct of cases, OR
    #   - Only 2 labels remain (can't drop further without losing all signal)
    #
    # --- Inputs ---
    # events  : pd.DataFrame — must have a 'bin' column with labels
    # min_pct : float — minimum fraction a label must represent (default 5%)
    #
    # --- Output ---
    # pd.DataFrame — same as events but with rare-label rows removed

    while True:
        df0 = events['bin'].value_counts(normalize=True)
        if df0.min() > min_pct or df0.shape[0] < 3:
            break
        rarest_label = df0.idxmin()
        print(f"Dropping label {rarest_label} ({df0.min()*100:.1f}% of cases)")
        events = events[events['bin'] != rarest_label]

    return events

File: ch03ts/meta_labeling.ts
// meta_labeling.ts — mirror of ch03/labeling/meta_labeling.py
// AFML Chapter 3, Sections 3.6-3.9, pages 50-54
//
// 📁 C:\ws\AFML\
// └── ch03_ts\
//     └── labeling_ts\
//         └── meta_labeling.ts   ← goes here

import { applyPtSlOnT1 } from './triple_barrier';
import type { ClosePrice, EventRow, BinLabel } from './triple_barrier';

// Re-export shared types so callers only need to import from one place
export type { ClosePrice, EventRow };

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/**
 * Output of getBinsMeta — final label with meta-labeling applied.
 * bin ∈ {0, 1} when side is provided (was primary model right or wrong?)
 * bin ∈ {-1, 0, 1} when side is absent (standard triple barrier)
 */
export interface MetaBinLabel {
  date: string;
  ret: number;        // actual return (sign-adjusted if side provided)
  bin: 1 | 0 | -1;   // label
}

// ---------------------------------------------------------------------------
// Helpers (shared with triple_barrier.ts)
// ---------------------------------------------------------------------------

/** Binary search: index of first date >= target in sortedDates. */
function searchSorted(sortedDates: string[], target: string): number {
  let lo = 0, hi = sortedDates.length;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (sortedDates[mid] < target) lo = mid + 1;
    else hi = mid;
  }
  return lo;
}

// ---------------------------------------------------------------------------
// getEventsMeta
// ---------------------------------------------------------------------------

/**
 * Expanded getEvents with Meta-Labeling — AFML Chapter 3, Snippet 3.6, page 51.
 *
 * Extension of getEvents() that adds support for a primary model's side
 * predictions. When side is provided, the barriers become asymmetric —
 * profit target and stop loss are applied relative to the predicted direction.
 *
 * Without side: symmetric barriers, assume long (+1) for all events.
 * With side:    barriers follow the primary model's direction.
 *               Primary long  (+1): upper = profit target, lower = stop loss
 *               Primary short (-1): upper = stop loss, lower = profit target
 *
 * @param close    - array of ClosePrice in chronological order
 * @param tEvents  - candidate entry dates (from CUSUM filter)
 * @param ptSl     - [pt_multiplier, sl_multiplier]
 * @param trgt     - array of { date, vol } from getDailyVol
 * @param minRet   - minimum target return to include an event
 * @param t1       - vertical barrier dates (null = no time limit)
 * @param side     - primary model's predicted side per event date (optional)
 *                   Map<date, +1 | -1>
 * @returns array of EventRow with t1 = first touch timestamp
 */
export function getEventsMeta(
  close: ClosePrice[],
  tEvents: string[],
  ptSl: [number, number],
  trgt: { date: string; vol: number }[],
  minRet: number,
  t1: { date: string; t1: string }[] | null = null,
  side: Map<string, number> | null = null
): EventRow[] {
  // Build vertical barrier map
  const t1Map = new Map<string, string | null>();
  if (t1 !== null) {
    for (const row of t1) t1Map.set(row.date, row.t1);
  }

  // Step 1: Filter trgt to event dates (bfill) and apply minimum return filter
  // Mirrors: trgt.reindex(t_events, method='bfill'); trgt[trgt > min_ret]
  const trgtDates = trgt.map(t => t.date);

  const events: EventRow[] = [];

  for (const date of tEvents) {
    // bfill: find next available vol at or after this date
    const idx = searchSorted(trgtDates, date);
    if (idx >= trgt.length) continue;
    const vol = trgt[idx].vol;
    if (vol <= minRet) continue; // minimum return filter

    const verticalBarrier = t1Map.get(date) ?? null;

    // Step 3: Determine side and barrier multipliers
    // No primary model → assume long (+1), symmetric barriers
    // Primary model provided → use its side, asymmetric barriers allowed
    const tradeSide = side !== null ? (side.get(date) ?? 1) : 1;

    events.push({
      date,
      t1:   verticalBarrier,
      trgt: vol,
      side: tradeSide,
    });
  }

  // Step 5: Find first barrier touch
  // No primary model → symmetric [pt, pt]; with model → [pt, sl] as given
  const ptSl_ = side === null
    ? [ptSl[0], ptSl[0]] as [number, number]  // symmetric
    : ptSl;                                    // asymmetric (primary model)

  const touches = applyPtSlOnT1(close, events, ptSl_);

  // Step 6: Take earliest touch across sl, pt, and vertical barrier
  const result: EventRow[] = [];

  for (let i = 0; i < events.length; i++) {
    const event = events[i];
    const touch = touches[i];

    const candidates: string[] = [];
    if (touch.sl !== null) candidates.push(touch.sl);
    if (touch.pt !== null) candidates.push(touch.pt);
    if (touch.t1 !== null) candidates.push(touch.t1);

    if (candidates.length === 0) continue;

    const firstTouch = candidates.reduce((a, b) => (a < b ? a : b));
    result.push({ ...event, t1: firstTouch });
  }

  return result;
}

// ---------------------------------------------------------------------------
// getBinsMeta
// ---------------------------------------------------------------------------

/**
 * Expanded getBins with Meta-Labeling — AFML Chapter 3, Snippet 3.7, pages 51-52.
 *
 * Assigns final labels with meta-labeling logic:
 *
 * Case 1 — No side (standard triple barrier):
 *   bin = sign(return) ∈ {-1, 0, +1} — direction of price move
 *
 * Case 2 — With side (meta-labeling):
 *   ret_adjusted = return × side  (flip sign for short trades)
 *   bin = 1 if primary model was correct (ret_adjusted > 0)
 *   bin = 0 if primary model was wrong   (ret_adjusted ≤ 0)
 *   bin ∈ {0, 1} — binary classification: right or wrong?
 *
 * Why binary?
 *   Separating direction (primary model) from correctness (meta-model) lets
 *   each model specialize. Binary classification is also easier for ML models.
 *
 * @param events    - output of getEventsMeta()
 * @param close     - array of ClosePrice in chronological order
 * @param hasSide   - true if events include primary model side predictions
 * @returns array of MetaBinLabel
 */
export function getBinsMeta(
  events: EventRow[],
  close: ClosePrice[],
  hasSide = false
): MetaBinLabel[] {
  const dates  = close.map(c => c.date);
  const prices = close.map(c => c.price);

  const results: MetaBinLabel[] = [];

  for (const event of events) {
    if (event.t1 === null) continue; // no exit date → skip

    // Entry price: first bar at or after entry date
    const entryIdx   = Math.min(searchSorted(dates, event.date), dates.length - 1);
    const entryPrice = prices[entryIdx];
    if (entryPrice === 0) continue;

    // Exit price: first bar at or after t1
    const exitIdx   = Math.min(searchSorted(dates, event.t1), dates.length - 1);
    const exitPrice = prices[exitIdx];

    let ret = exitPrice / entryPrice - 1;

    let bin: 1 | 0 | -1;

    if (hasSide) {
      // Meta-labeling case: adjust return by primary model's side
      // Long (+1) and price up → positive → correct
      // Short (-1) and price down → negative × -1 → positive → correct
      ret = ret * event.side;

      // bin = 1 if correct (ret > 0), 0 if wrong (ret ≤ 0)
      // ret === 0 (vertical barrier, neutral) counts as wrong
      bin = ret > 0 ? 1 : 0;
    } else {
      // Standard case: label by direction of price move
      bin = ret > 0 ? 1 : ret < 0 ? -1 : 0;
    }

    results.push({ date: event.date, ret, bin });
  }

  return results;
}

// ---------------------------------------------------------------------------
// dropLabels
// ---------------------------------------------------------------------------

/**
 * Dropping Under-Populated Labels — AFML Chapter 3, Snippet 3.8, page 53.
 *
 * Recursively removes rows with extremely rare labels until either:
 *   - All remaining labels appear in at least minPct of cases, OR
 *   - Only 2 labels remain (can't drop further without losing all signal)
 *
 * Why?
 *   If 95% of labels are +1 and only 5% are -1, a model can hit 95% accuracy
 *   by always predicting +1 — without learning anything useful. Removing
 *   rare labels forces the model to actually learn the harder cases.
 *
 * @param labels  - array of MetaBinLabel (must have bin field)
 * @param minPct  - minimum fraction a label must represent (default 0.05 = 5%)
 * @returns filtered array with rare-label rows removed
 */
export function dropLabels(
  labels: MetaBinLabel[],
  minPct = 0.05
): MetaBinLabel[] {
  let current = [...labels];

  while (true) {
    const total = current.length;
    if (total === 0) break;

    // Count how often each label appears
    // Mirrors: events['bin'].value_counts(normalize=True)
    const counts = new Map<number, number>();
    for (const row of current) {
      counts.set(row.bin, (counts.get(row.bin) ?? 0) + 1);
    }

    // Convert to fractions
    const fractions = new Map<number, number>();
    for (const [label, count] of counts) {
      fractions.set(label, count / total);
    }

    const minFrac  = Math.min(...fractions.values());
    const numLabels = fractions.size;

    // Stop if all labels meet the threshold or only 2 labels remain
    if (minFrac > minPct || numLabels < 3) break;

    // Find and drop the rarest label
    let rarestLabel = -999;
    for (const [label, frac] of fractions) {
      if (frac === minFrac) { rarestLabel = label; break; }
    }

    console.log(
      `Dropping label ${rarestLabel} (${(minFrac * 100).toFixed(1)}% of cases)`
    );

    current = current.filter(row => row.bin !== rarestLabel);
  }

  return current;
}
